apps/ejabberd/src/ejabberd_s2s.erl

```diff
@@ -547,33 +547,39 @@ update_tables() ->
     end.
 
 %% Check if host is in blacklist or white list
+%% first see if it is a subdomain of something we know
 allow_host(MyServer, S2SHost) ->
     Hosts = ?MYHOSTS,
     case lists:dropwhile(
            fun(ParentDomain) ->
                    not lists:member(ParentDomain, Hosts)
            end, parent_domains(MyServer)) of
         [MyHost|_] ->
-            allow_host1(MyHost, S2SHost);
+            allow_host_local_option(MyHost, S2SHost);
         [] ->
-            allow_host1(MyServer, S2SHost)
+            allow_host_local_option(MyServer, S2SHost)
     end.
 
-allow_host1(MyHost, S2SHost) ->
-    case ejabberd_config:get_local_option({{s2s_host, S2SHost}, MyHost}) of
+allow_host_local_option(MyHost, S2SHost) ->
+    LocalDecision = ejabberd_config:get_local_option({{s2s_host, S2SHost}, MyHost}),
+    allow_host_default_policy(LocalDecision, MyHost, S2SHost).
+
+allow_host_default_policy(deny, _, _) ->
+    false;
+allow_host_default_policy(allow, _, _) ->
+    true;
+allow_host_default_policy(_, MyHost, S2SHost) ->
+    DefaultDecision =  ejabberd_config:get_local_option({s2s_default_policy, MyHost}),
+    allow_host_hook(DefaultDecision, MyHost, S2SHost).
+
+allow_host_hook(deny, _, _) ->
```

Comment:
Good point, in fact in this case it is not a 'default' policy but something enforced. It is the original implementation, I'll try to change it.